src/transpose_y_to_z.f90

```diff
@@ -94,57 +93,77 @@ subroutine transpose_y_to_z_real(src, dst, decomp, wk1, wk2)
       ! define receive buffer
 #ifdef EVEN
       if (decomp%even) then
+#   if defined(_GPU)
+         call MPI_ALLTOALL(wk1, decomp%y2count, real_type, &
+                           wk2, decomp%z2count, real_type, &
+                           DECOMP_2D_COMM_ROW, ierror)
+#   else
          call MPI_ALLTOALL(wk1, decomp%y2count, real_type, &
                            dst, decomp%z2count, real_type, &
                            DECOMP_2D_COMM_ROW, ierror)
+#   endif
       else
          call MPI_ALLTOALL(wk1, decomp%y2count, real_type, &
                            wk2, decomp%z2count, real_type, &
                            DECOMP_2D_COMM_ROW, ierror)
       end if
       if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_ALLTOALL")
-
-#elif defined(_NCCL)
+#else
+    ! MPI_ALLTOALLV branch => NO EVEN
+#   if defined(_GPU)
+#     if defined(_NCCL)
       call decomp_2d_nccl_send_recv_row(wk2, &
                                         wk1, &
                                         decomp%y2disp, &
                                         decomp%y2cnts, &
                                         decomp%z2disp, &
                                         decomp%z2cnts, &
                                         dims(2))
-
-#elif defined(_GPU)
+#     else 
+      ! CUDA aware MPI
       call MPI_ALLTOALLV(wk1, decomp%y2cnts, decomp%y2disp, real_type, &
                          wk2, decomp%z2cnts, decomp%z2disp, real_type, &
                          DECOMP_2D_COMM_ROW, ierror)
       if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_ALLTOALLV")
 
-#else
+#     endif
+#   else
+      ! Pure MPI => direct to DST
       associate (wk => wk2)
       end associate
       call MPI_ALLTOALLV(wk1, decomp%y2cnts, decomp%y2disp, real_type, &
                          dst, decomp%z2cnts, decomp%z2disp, real_type, &
                          DECOMP_2D_COMM_ROW, ierror)
       if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_ALLTOALLV")
+
+#   endif
 #endif
 
       ! rearrange receive buffer
 #ifdef EVEN
-      if (.not. decomp%even) then
+      if (decomp%even) then
+#   if defined(_GPU)
+        !If one of the array in cuda call is not device we need to add acc host_data
+        !$acc host_data use_device(dst)
+        istat = cudaMemcpy(dst, wk2, d1 * d2 * d3, cudaMemcpyDeviceToDevice)
```

Comment:
If `dst` can be used to receive the MPI communication, then the line here is needed only for NCCL